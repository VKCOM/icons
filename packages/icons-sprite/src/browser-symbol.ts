import { parse } from './utils/parse';

interface SpriteSymbolOption {
  content: string;
}

export class BrowserSymbol {
  protected isMounted: boolean = false;
  node: Element;

  constructor({ content }: SpriteSymbolOption) {
    this.node = parse(content);
  }

  get id(): string {
    return this.node.id;
  }

  static createFromExistingNode(node: Element): BrowserSymbol {
    const symbol = new BrowserSymbol({
      content: '',
    });
    symbol.node = node;

    return symbol;
  }

  mount(target: Node) {
    if (this.isMounted) {
      return this.node;
    }

    this.isMounted = true;

    target.appendChild(this.node);

    return this.node;
  }

  unmount() {
    if (this.node && this.node.parentNode) {
      this.node.parentNode.removeChild(this.node);
      this.isMounted = false;
    }
  }
}
