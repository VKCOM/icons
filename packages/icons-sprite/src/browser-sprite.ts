import { BrowserSymbol } from './browser-symbol';

const namespaceURI = 'http://www.w3.org/2000/svg';

export interface SpriteConfig {
  attrs?: Record<string, string>;
}

export class BrowserSprite {
  private symbols: Map<string, BrowserSymbol> = new Map();
  private config: SpriteConfig = {
    attrs: {
      'xmlns': namespaceURI,
      'xmlns:xlink': 'http://www.w3.org/1999/xlink',
      'style': 'position:absolute;width:0;height:0',
      'aria-hidden': 'true',
    },
  };

  node: Element | null = null;

  constructor(config: SpriteConfig = {}) {
    Object.assign(this.config.attrs!, config.attrs);
  }

  private push(symbol: BrowserSymbol) {
    const { symbols } = this;
    const existing = symbols.has(symbol.id);

    symbols.set(symbol.id, symbol);

    return !existing;
  }

  add(symbol: BrowserSymbol) {
    const isNewSymbol = this.push(symbol);

    if (!!this.node && isNewSymbol) {
      symbol.mount(this.node);
    }

    return isNewSymbol;
  }

  attach(target: Element) {
    if (!!this.node) {
      return;
    }

    this.node = target;

    // Already added symbols needs to be mounted
    this.symbols.forEach((symbol) => {
      symbol.mount(target);
    });

    // Create symbols from existing DOM nodes, add and mount them
    target.querySelectorAll('symbol').forEach((symbolNode: Element) => {
      const symbol = BrowserSymbol.createFromExistingNode(symbolNode);
      this.add(symbol);
    });
  }

  mount(target: Node, prepend = false) {
    if (!!this.node) {
      return this.node;
    }

    this.node = this.render();

    if (prepend && target.childNodes[0]) {
      target.insertBefore(this.node, target.firstChild);
    } else {
      target.appendChild(this.node);
    }

    return this.node;
  }

  private render() {
    const el = document.createElementNS(namespaceURI, 'svg');

    Object.entries(this.config.attrs!).forEach((entry) => el.setAttribute(entry[0], entry[1]));
    this.symbols.forEach((symbol) => el.appendChild(symbol.node));

    return el;
  }

  unmount() {
    this.node && this.node.parentNode && this.node.parentNode.removeChild(this.node);
  }
}
