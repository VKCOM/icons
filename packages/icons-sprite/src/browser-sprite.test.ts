import { describe, expect, test, afterEach } from '@jest/globals';
import { BrowserSprite } from './browser-sprite';
import { BrowserSymbol } from './browser-symbol';

afterEach(() => {
  document.body.innerHTML = '';
});

describe(BrowserSprite, () => {
  test('mount && unmount', () => {
    document.body.innerHTML = '<div id="block"></div>';
    const sprite = new BrowserSprite({
      attrs: {
        id: 'sprite',
      },
    });

    sprite.add(new BrowserSymbol({ content: `<symbol id="symbol1"></symbol>` }));
    const node = sprite.mount(document.body);
    sprite.add(new BrowserSymbol({ content: `<symbol id="symbol2"></symbol>` }));

    expect(document.getElementById('sprite')).not.toBeNull();
    expect(document.getElementById('symbol1')).not.toBeNull();
    expect(document.getElementById('symbol2')).not.toBeNull();
    expect(document.getElementById('block')).not.toBeNull();
    expect(document.body.firstElementChild!.id).toBe('block');
    expect(document.body).toMatchSnapshot('mount');
    expect(sprite.mount(document.body)).toBe(node);

    sprite.unmount();

    expect(document.getElementById('sprite')).toBeNull();
    expect(document.getElementById('symbol1')).toBeNull();
    expect(document.getElementById('symbol2')).toBeNull();
    expect(document.getElementById('block')).not.toBeNull();
    expect(document.body).toMatchSnapshot('unmount');
  });

  test('mount && unmount prepend', () => {
    document.body.innerHTML = '<div id="block"></div>';

    const sprite = new BrowserSprite({
      attrs: {
        id: 'sprite',
      },
    });

    sprite.mount(document.body, true);
    sprite.add(new BrowserSymbol({ content: `<symbol id="symbol"></symbol>` }));

    expect(document.getElementById('sprite')).not.toBeNull();
    expect(document.getElementById('symbol')).not.toBeNull();
    expect(document.getElementById('block')).not.toBeNull();
    expect(document.body.firstElementChild!.id).toBe('sprite');
    expect(document.body).toMatchSnapshot('mount');

    sprite.unmount();

    expect(document.getElementById('sprite')).toBeNull();
    expect(document.getElementById('symbol')).toBeNull();
    expect(document.getElementById('block')).not.toBeNull();
    expect(document.body).toMatchSnapshot('unmount');
  });

  test('attach', () => {
    const sprite = new BrowserSprite({
      attrs: {
        id: 'sprite',
      },
    });
    const node = sprite.mount(document.body);
    sprite.add(new BrowserSymbol({ content: `<symbol id="symbol1"></symbol>` }));

    const sprite2 = new BrowserSprite();
    sprite2.add(new BrowserSymbol({ content: `<symbol id="symbol2"></symbol>` }));
    sprite2.attach(node);

    expect(document.getElementById('sprite')).not.toBeNull();
    expect(document.getElementById('symbol1')).not.toBeNull();
    expect(document.getElementById('symbol2')).not.toBeNull();
    expect(document.body).toMatchSnapshot();

    sprite2.attach(node);
  });
});
