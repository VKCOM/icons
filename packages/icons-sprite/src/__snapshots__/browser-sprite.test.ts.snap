// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserSprite attach 1`] = `
<body>
  <svg
    aria-hidden="true"
    id="sprite"
    style="position:absolute;width:0;height:0"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <symbol
      id="symbol2"
    />
    <symbol
      id="symbol1"
    />
  </svg>
</body>
`;

exports[`BrowserSprite mount && unmount prepend: mount 1`] = `
<body>
  <svg
    aria-hidden="true"
    id="sprite"
    style="position:absolute;width:0;height:0"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <symbol
      id="symbol"
    />
  </svg>
  <div
    id="block"
  />
</body>
`;

exports[`BrowserSprite mount && unmount prepend: unmount 1`] = `
<body>
  <div
    id="block"
  />
</body>
`;

exports[`BrowserSprite mount && unmount: mount 1`] = `
<body>
  <div
    id="block"
  />
  <svg
    aria-hidden="true"
    id="sprite"
    style="position:absolute;width:0;height:0"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <symbol
      id="symbol1"
    />
    <symbol
      id="symbol2"
    />
  </svg>
</body>
`;

exports[`BrowserSprite mount && unmount: unmount 1`] = `
<body>
  <div
    id="block"
  />
</body>
`;
